internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+	}
+	if exists {
+		return LoadKey(keyDir)
+	}
+	return GenerateKey(keyDir)
+}
+
+// GenerateKey generates a new identity key and persists it to disk. This
+// function should only ever be called on the first boot otherwise the
+// existing identity will be overwritten.
+//
+// This function is equivalent to running:
+//
+//	openssl genpkey -algorithm Ed25519 -out key.pem
+func GenerateKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	err := key.createDir()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create identity directory: %w", err)
+	}
+	err = key.newEd25519()
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate identity key: %w", err)
+	}
+	err = key.save()
+	if err != nil {
+		return nil, fmt.Errorf("cannot save identity key: %w", err)
+	}
+	return key, nil
+}
+
+// LoadKey loads an existing identity key from disk.
+func LoadKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	// Load from disk.
+	err := key.load()
+	if err != nil {
+		return nil, fmt.Errorf("cannot load identity key: %w", err)
+	}
+	return key, nil
+}
+
+// createDir verifies the directory layout and permissions, and creates
+// the leaf element of the key driectory if it does not yet exist.
+func (k *IDKey) createDir() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(k.keyDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(k.keyDir, 0o700)
+}
+
+// Public implements the crypto.Signer.
+func (k *IDKey) Public() crypto.PublicKey {
+	signer := k.key.(crypto.Signer)
+	return signer.Public()
+}
+
+// Sign implements the crypto.Signer.
+func (k *IDKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	signer := k.key.(crypto.Signer)
+	return signer.Sign(rand, digest, opts)
+}
+
+// newEd25519 generates a new ed25519 key.
+func (k *IDKey) newEd25519() error {
+	var err error
+	_, k.key, err = ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// load loads the private identity key from storage.
+func (k *IDKey) load() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil || !exists {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
+	}
+	err = expectPermission(k.keyDir, 0o700)
+	if err != nil {
+		return err
+	}
+	pemPath := filepath.Join(k.keyDir, identityKeyFile)
+	err = expectPermission(pemPath, 0o600)
+	if err != nil {
+		return err
+	}
+	// Load the key.
+	pemData, err := os.ReadFile(pemPath)
+	if err != nil {
+		return err
+	}
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			break
+		}
+		switch block.Type {
+		case "PRIVATE KEY":
+			k.key, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown PEM block %q in file %q", block.Type, pemPath)
+		}
+	}
```

Comment:
This whole loop is unnecessary if we expect to have a single private key block in the file. If we want to support multiple blocks, we should definitely ignore the blocks that are not private keys instead of returning an error early.

Reply:
I removed the whole loop and simplified it to expect only the single "PRIVATE KEY" block. Resolving.